src/group_theory/is_free_group.lean

```diff
@@ -1,175 +1,132 @@
 
-/-- Any free group is isomorphic to "the" free group. -/
-@[simps] def to_free_group : G ≃* free_group (generators G) :=
-{ to_fun := lift' free_group.of,
-  inv_fun := free_group.lift of,
-  left_inv :=
-    suffices (free_group.lift of).comp (lift' free_group.of) = monoid_hom.id G,
-    from monoid_hom.congr_fun this,
-    by { ext, simp },
-  right_inv :=
-    suffices
-      (lift' free_group.of).comp (free_group.lift of) = monoid_hom.id (free_group (generators G)),
-    from monoid_hom.congr_fun this,
-    by { ext, simp },
-  map_mul' := (lift' free_group.of).map_mul }
+variables {H : Type*} [group H]
 
-variable {G}
+/-- The equivalence between functions on the generators and group homomorphisms from a free group
+given by those generators. -/
+@[simps symm_apply]
+def lift : (generators G → H) ≃ (G →* H) :=
+free_group.lift.trans
+  { to_fun := λ f, f.comp (@is_free_group.mul_equiv G _ _).symm.to_monoid_hom,
+    inv_fun := λ f, f.comp (@is_free_group.mul_equiv G _ _).to_monoid_hom,
+    left_inv := λ f, by { ext, simp, },
+    right_inv := λ f, by { ext, simp, }, }
 
-private lemma lift_right_inv_aux (F : G →* X) :
-  free_group.lift.symm (F.comp (to_free_group G).symm.to_monoid_hom) = F ∘ of :=
-by { ext, simp }
+@[simp] lemma lift'_eq_free_group_lift {A : Type u} :
+  (@lift (free_group A) _ _ H _) = free_group.lift := rfl
 
-/-- A universe-polymorphic version of `is_free_group.lift'`. -/
-@[simps symm_apply]
-def lift : (generators G → X) ≃ (G →* X) :=
-{ to_fun := λ f, (free_group.lift f).comp (to_free_group G).to_monoid_hom,
-  inv_fun := λ F, F ∘ of,
-  left_inv := λ f, free_group.lift.injective begin
-    ext x,
-    simp,
-  end,
-  right_inv := λ F, begin
-    dsimp,
-    rw ←lift_right_inv_aux,
-    simp only [equiv.apply_symm_apply],
-    ext x,
-    dsimp only [monoid_hom.comp_apply, mul_equiv.coe_to_monoid_hom],
-    rw mul_equiv.symm_apply_apply,
-  end}
-
-@[ext]
-lemma ext_hom ⦃f g : G →* X⦄ (h : ∀ a, f (of a) = g (of a)) :
-  f = g :=
-is_free_group.lift.symm.injective $ funext h
-
-@[simp] lemma lift_of (f : generators G → X) (a : generators G) : (lift f) (of a) = f a :=
+@[simp] lemma lift_of (f : generators G → H) (a : generators G) : lift f (of a) = f a :=
 congr_fun (lift.symm_apply_apply f) a
 
-@[simp] lemma lift_eq_free_group_lift {A : Type u} :
-  (@lift (free_group A) H _ _ _) = free_group.lift :=
-begin
-  -- TODO: `apply equiv.symm_bijective.injective`,
-  rw [←free_group.lift.symm_symm, ←(@lift (free_group A) H _ _ _).symm_symm],
-  congr' 1,
-  ext,
-  refl,
-end
-
-/-- A universe-polymorphic version of `unique_lift`. -/
-lemma unique_lift {X : Type w} [group X] (f : generators G → X) :
-  ∃! F : G →* X, ∀ a, F (of a) = f a :=
-begin
-  have := lift.symm.bijective.exists_unique f,
-  simp_rw function.funext_iff at this,
-  exact this,
-end
+@[simp] lemma lift_symm_of (f : G →* H) (a : generators G) : (lift.symm f) a = f (of a) :=
+rfl
```

Comment:
This lemma should be removed, the name isn't accurate anyway and the one with the correct name, `is_free_group.lift_symm_apply`, already exists

Reply:
Ah, looks like we raced. In the end I did include that lemma, with a correct name, as I would find that helpful in the docs. I removed `is_free_group.lift_symm_apply`.